File: 프로그래머스/LV2/코딩테스트 연습/Heap/더 맵게/더 맵게.py
from heapq import heapify, heappush, heappop

def cut_pos(scov, k):
    for i in range(len(scov)):
        if scov[i] >= k:
            break
    return i + 1

def solution(scoville, K):

    answer = 0
    scoville.sort()
    scoville = scoville[:cut_pos(scoville, K)]
    heapify(scoville)
    heap_size = len(scoville)

    while heap_size > 1:
        temp = scoville[0]
        heappop(scoville)
        temp += scoville[0] * 2
        heappop(scoville)

        heappush(scoville, temp)

        answer += 1
        heap_size -= 1

        if scoville[0] >= K:
            return answer

    if scoville[0] < K:
        answer = -1
    return answer

File: 프로그래머스/LV2/코딩테스트 연습/Heap/더 맵게/더 맵게.cpp
#include <vector>
#include <queue>

using namespace std;
typedef priority_queue<int, vector<int>, greater<int>> HEAP;

int solution(vector<int> scoville, int K)
{
    int answer = 0;
    int temp = 0;
	
    HEAP heap(scoville.begin(), scoville.end());

    int heap_size = scoville.size();

    while (heap_size > 1)
    {
        temp = heap.top();
        heap.pop();
        heap.push(temp + heap.top() * 2);
        heap.pop();

        answer++;
        heap_size--;

        if (heap.top() >= K)
            return answer;
    }
    if (heap.top() < K)
        answer = -1;
    return answer;
}